### PR TITLE
Feature/sticky resource pool

### PR DESCRIPTION
**Pull request by steveshogren**
Should match what is in the feature request. I added the check to allow different threads to each have their own resource in the pool, so they are segregated by parameter and ThreadId.

The entry points are ```pool``` and ```poolWithoutGC```. The regular ```pool``` will spawn a second thread that checks for expired resources every second. The ```poolWithoutGC``` will not spawn that thread, so the cache will never be cleaned.

The one gotcha right now is what happens when several requests are made to the same resource from the same thread without any of the requests being released. Right now that will cause undefined behavior as each "release" will overwrite the previous releases entry in the pool. I am not sure exactly what should happen in that case. 

It should be fairly well tested, which might account for a few oddities in the function signatures (like passing around a getter for the ThreadId). I am not sure how idiomatic that is in regular Haskell, but it struck me as better than manually testing everything.

Feel free to comment/reject, I am happy to change whatever!

### CODE REVIEW

File: node/src/Unison/Runtime/ResourcePool.hs [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+import Data.Time (UTCTime, getCurrentTime, addUTCTime)
+import qualified Control.Concurrent as CC
+
+-- acquire returns the resource, and the cleanup action ("finalizer") for that resource
+data Pool p r = Pool { acquire :: p -> Int -> IO (r, IO ()) }
+
+type ResourceKey p = (p,CC.ThreadId)
+type Cache p r = MVar.MVar (M.Map (ResourceKey p) (r, UTCTime, IO ()))
+
+addResourceToMap :: (Ord p) =>  (r -> IO()) -> Cache p r -> p -> r -> Int -> Int -> IO(CC.ThreadId) -> IO ()
```
Comment (pchiusano) [inline node/src/Unison/Runtime/ResourcePool.hs]: Minor formatting stuff -

* If just one constraint, leave off the parens - `Ord p => ...`.
* `IO ()` and `IO CC.ThreadId`, type application should be formatted same as function application - space after the type constructor, no parens around the args unless necessary

File: node/src/Unison/Runtime/ResourcePool.hs [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+addResourceToMap releaser mVarCache p r wait maxPoolSize getThread = do
+  now <- getCurrentTime
+  threadId <- getThread
+  let expiry = addUTCTime (fromIntegral wait) now
+  cacheMap <- MVar.takeMVar mVarCache
```
Comment (pchiusano) [inline node/src/Unison/Runtime/ResourcePool.hs]: Use `modifyMVar_` - lets you safely modify an `MVar`, putting back the original value if an exception occurs. The code you have here is missing exception handling, so if `release r` bombs the `MVar` is left empty.

File: node/src/Unison/Runtime/ResourcePool.hs [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+                                      case M.lookup k cacheMap of
+                                          Just (_, expiry, releaser) ->
+                                            if expiry < now then (k, releaser) : knrs
+                                            else knrs
+                                          Nothing -> knrs) [] cacheMap
+      newMap = foldr (\(k,_) m -> M.delete k m) cacheMap keysNReleasers
+  (sequence $ map snd keysNReleasers)
+    >> MVar.putMVar mVarCache newMap
+
+cleanCacheLoop :: Ord p => Cache p r -> IO b
```
Comment (pchiusano) [inline node/src/Unison/Runtime/ResourcePool.hs]: So, I think this approach is reasonable, but what isn't ideal is that it's touching every key in the pool on each wakeup. You could optimize this by scheduling the deletion from the pool at the time the resource is acquired (or refreshed). As in, you have a `Chan (p, UTCTime)` of key, expiration time. Then you have your 'reaper thread' which just pulls from this channel in a loop and sleeps until the next expiration time, touching only that key, no need to traverse the whole map.

https://hackage.haskell.org/package/base-4.8.2.0/docs/Control-Concurrent-Chan.html

Also, I think it may be better to use a concurrent map - https://hackage.haskell.org/package/ctrie-0.1.0.3/docs/Control-Concurrent-Map.html rather than an `MVar (Map k v)`. Should be less contention, and feels nicer.

File: node/tests/Unison/Test/ResourcePool.hs [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+  didAcquire <- loadState ts "testacquires"
+  didRelease <- loadState ts "testreleases"
+
+  assertEqual "the correct resource is returned" "p1r" r1
+    >> assertEqual "the correct resource is returned" "p1r" r2
+    >> assertEqual "the correct resource is returned" "p1r" r3
+    >> assertEqual "r is acquired twice" "p1rp1r" didAcquire
+
+tests :: TestTree
+tests = testGroup "Doc"
```
Comment (pchiusano) [inline node/tests/Unison/Test/ResourcePool.hs]: `testGroup "ResourcePool"` am guessing this was copy/paste typo :)

File: node/src/Unison/Runtime/ResourcePool.hs (hunk 2 above)
Comment (pchiusano) [inline node/src/Unison/Runtime/ResourcePool.hs]: Okay, ignore this comment, I want to just kill the use of `MVar`.

**Comment by pchiusano** [comment]
Hey @steveshogren thanks for this. I think the big change I'd like to try is this one:

https://github.com/unisonweb/unison/pull/38/files#r55942425

So getting rid of `MVar` in favor of using a concurrent map, and then having the reaper thread read from a `TChan` or `TQueue` and just (fast path) touch a single key on each wakeup. The one trick to doing this is that when the reaper thread wakes up, it should read from the `TChan` (using `tryPeekTChan` I think) for as long as there are expired keys). Reason is that we don't have any guarantees about how fair GHC will be in thread scheduling of the reaper, and so we want it to catch up processing all expired resources when it does get scheduled.

* [`TChan`](https://hackage.haskell.org/package/stm-2.4.4.1/docs/Control-Concurrent-STM-TChan.html)
* [`TQueue`](https://hackage.haskell.org/package/stm-2.4/docs/Control-Concurrent-STM-TQueue.html) but `TQueue` claims to be faster. Both live in `STM` but you can convert that to `IO` using [`atomically`](https://hackage.haskell.org/package/stm-2.4/docs/Control-Monad-STM.html#t:STM).

**Comment by steveshogren** [comment]
Sounds great, thanks for taking the time to carefully review it! I'll get on those changes and have it fixed up this week. Also I'll watch out for those style changes going forward.

**Comment by pchiusano** [comment]
Cool. Take your time and have fun! :) We are not blocked on this right now
but it will definitely be useful.

**Comment by steveshogren** [comment]
Getting rid of MVar for Concurrent.Map means we cannot count the size of the map, so we'd need to still use an MVar to hold the count of the map size in it. Perhaps considering that, it makes more sense to stay with MVar?

**Comment by pchiusano** [comment]
Hmm yeah that is a problem. I'm surprised concurrent map doesn't track that.

I would still use it though and just track the size separately in a var.
I'd use a TVar or something STM based since retries are fine, we just need
atomic increments and decrements. MVar probably overkill and will be slower
for that.

**Comment by pchiusano** [comment]
Actually it looks like IORef is the way to go, and it has a CAS based
modify operation:
http://stackoverflow.com/questions/15439966/when-why-use-an-mvar-over-a-tvar

**Comment by steveshogren** [comment]
@pchiusano I think I've got it tweaked to use the queue. I put a lot of detail in the https://github.com/unisonweb/unison/pull/38/commits/32c30656531ae1afca8828e1011a7ce5b628da76 commit, it helped me conceptualize the workflow.

File: node/src/Unison/Runtime/ResourcePool.hs [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+  m <- M.lookup (p,threadId) pool
+  r <- case m of
+        Just (r, _, _) -> M.delete (p,threadId) pool
+          >> IORef.atomicModifyIORef poolSizeR decRef
+          >> return r
+        Nothing -> acquire p
+  return (r, (addResourceToMap release pool p r wait maxPoolSize getThread poolSizeR queue))
+
+cleanPool :: (Ord p, H.Hashable p) => Cache p r -> ReaperQueue p -> IO ()
+cleanPool pool queue = do
```
Comment (pchiusano) [inline node/src/Unison/Runtime/ResourcePool.hs]: This isn't quite what I had in mind. This is polling the `ReaperQueue` periodically. But we can make it so there's no polling. I want the logic to be:

1. Try reading from the queue via `tryPeekTQueue`. If `Nothing`, then do a `readTQueue`, which will cause the reaper thread to sleep until a msg is available.
2. Okay, now we have a msg available (either from the peek or from the read), compare the `expiry` to `now`, then `threadDelay` for `expiry - now` (might be negative, in which case don't sleep). After this sleep, we're guaranteed this msg is expired, so check the `cache` to see if its safe to invoke the finalizer for that key. Then go to 1)

That is the loop. Notice there's no polling, we just sleep for however long is needed to reach the next time when a key might be expired.

File: node/src/Unison/Runtime/ResourcePool.hs [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+          Just (_, expiry, releaser) ->
+            if expiry < now then
+              M.delete key pool >> releaser
+                 >> cleanPool pool queue
+            else (STM.atomically $ TQ.writeTQueue queue msg)
+          Nothing -> return ()
+    Nothing -> return ()
+
+cleanPoolLoop :: (Ord p, H.Hashable p) => Cache p r -> ReaperQueue p -> IO b
+cleanPoolLoop mVarCache q =
```
Comment (pchiusano) [inline node/src/Unison/Runtime/ResourcePool.hs]: I'd just call the var `cache`, since it doesn't use `MVar` anymore. :)

File: node/src/Unison/Runtime/ResourcePool.hs [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+
+import Data.Time (UTCTime, getCurrentTime, addUTCTime)
+import qualified Control.Concurrent as CC
+import qualified Control.Concurrent.Map as M
+import qualified Control.Concurrent.STM.TQueue as TQ
+import qualified Data.Hashable as H
+import qualified Control.Monad.STM as STM
+import qualified Data.IORef as IORef
+
+-- acquire returns the resource, and the cleanup action ("finalizer") for that resource
```
Comment (pchiusano) [inline node/src/Unison/Runtime/ResourcePool.hs]: We could use a comment on the meaning of the `Int` here.

File: node/src/Unison/Runtime/ResourcePool.hs [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+        Nothing -> acquire p
+  return (r, (addResourceToMap release pool p r wait maxPoolSize getThread poolSizeR queue))
+
+cleanPool :: (Ord p, H.Hashable p) => Cache p r -> ReaperQueue p -> IO ()
+cleanPool pool queue = do
+  next <- STM.atomically $ TQ.tryPeekTQueue queue
+  now <- getCurrentTime
+  case next of
+    Just key -> do
+        value <- M.lookup key pool
```
Comment (pchiusano) [inline node/src/Unison/Runtime/ResourcePool.hs]: Minor formatting issue - just indent 2 spaces.

**Comment by pchiusano** [comment]
@steveshogren good work, this is very close. [This is main thing I'd like to see addressed](https://github.com/unisonweb/unison/pull/38/files#r57505418) before merging. 

I have some other ideas for projects after this one that will build on this work.

File: node/src/Unison/Runtime/ResourcePool.hs (hunk 5 above)
Comment (steveshogren) [inline node/src/Unison/Runtime/ResourcePool.hs]: Doesn't this break the ability for the pool to have configurable wait times in the cache? Now, if two resources are acquired, and the first releases first with a wait time of 10 seconds, and the second releases with a wait time of 1 second, they will both release after 10 seconds.

**Comment by steveshogren** [comment]
@pchiusano I updated your comment with my reply.

File: node/src/Unison/Runtime/ResourcePool.hs (hunk 5 above)
Comment (pchiusano) [inline node/src/Unison/Runtime/ResourcePool.hs]: Ah, yes, good point. :) I can think of some ways around that but they seem overly complicated. Let's just stick with the polling for now, we can always revisit later, once we are actually using this!

**Comment by steveshogren** [comment]
@pchiusano Is there anything else you wanted to see addressed before this can be merged?

File: node/src/Unison/Runtime/ResourcePool.hs (hunk 5 above)
Comment (pchiusano) [inline node/src/Unison/Runtime/ResourcePool.hs]: Okay, so I realized that the current polling logic is not correct either. The issue is that if we are allowing separate TTLs for each resource, we can't be guaranteed that the thing at the front of the queue is the msg closest to expiration. Thus, using that as the stopping condition for the drain loop is not correct. There might be an expired resource several messages behind the one you gave up on. Tell me if I am missing anything with my reasoning here.

One way to fix this would be to drain items from the queue into a heap, ordered by expiration time, but this is overkill.

The other thing is I don't think the reaper thread will ever be GC'd, even when the pool is no longer being referenced by anyone else. The issue is that it *always* wakes up every second. That is kind of bad. It would be better to kill that thread when the pool (and hence the `TQueue`) is no longer being referenced by anyone else. I have [a question out on that][so].

[so]: http://stackoverflow.com/questions/36310299/will-a-haskell-thread-be-gcd-if-blocked-indefinitely-on-an-mvar-chan-tqueue-rea

I think having separate TTL for each resource is overkill. Just have a single TTL for all resources in the `Pool`. If you really need multiple settings, just use different pools. For this use case, I think [the logic I gave above works](https://github.com/unisonweb/unison/pull/38/files#r57505418), though I want to get a definite answer on [whether the reaper thread will get GC'd if it is blocked indefinitely][so].

Comment (daniel-kun) [inline node/src/Unison/Runtime/ResourcePool.hs]: Aren't there tests for these circumstances to be guaranteed to be handled correctly?

Comment (pchiusano) [inline node/src/Unison/Runtime/ResourcePool.hs]: Okay, yeah, based on responses to the SO question and discussions with a couple other folks, the thread should be killed with a `*BlockedIndefinitely` exception so the approach I outlined above should work. You can do a quick sanity check that allocating new pools in a loop and then immediately discarding them doesn't leak memory (in the form of zombie threads).

**Comment by steveshogren** [comment]
@pchiusano I updated this with the new logic.

File: node/src/Unison/Runtime/ResourcePool.hs [hunk 9]
```diff
@@ -36,8 +35,8 @@ addResourceToMap release pool p r wait maxPoolSize getThread poolSizeR queue = d
       >> STM.atomically (TQ.writeTQueue queue key)
     else release r
 
-_acquire :: (Ord p, H.Hashable p) => (p -> IO r) -> (r -> IO ()) -> Cache p r -> MaxPoolSize -> IORef.IORef Int -> ReaperQueue p -> IO CC.ThreadId -> p -> Int -> IO (r, IO ())
-_acquire acquire release pool maxPoolSize poolSizeR queue getThread p wait = do
+_acquire :: (Ord p, H.Hashable p) => (p -> IO r) -> (r -> IO ()) -> Cache p r -> Int -> MaxPoolSize -> IORef.IORef Int -> ReaperQueue p -> IO CC.ThreadId -> p -> IO (r, IO ())
+_acquire acquire release pool wait maxPoolSize poolSizeR queue getThread p = do
   threadId <- getThread
   m <- M.lookup (p,threadId) pool
```
Comment (pchiusano) [inline node/src/Unison/Runtime/ResourcePool.hs]: @steveshogren Okay, so I'm not sure that keying the map on `(p, ThreadId)` is ideal, sorry I didn't spot this sooner. That means that if a resource is first acquired by thread A, then released, then thread B comes along before expiration and attempts to acquire the same resource, it won't get the cached one. So to get the benefit of using the pool, you need to acquire all your resources on the same thread. I think that is little too fragile.

I get that we need some way of preventing the resource from being simultaneously leased out to two different threads. Can you come up with a solution for that? Might be as simple as:

```Haskell
type IsLeased = Bool
type Cache p r = M.Map p (r, IsLeased, UTCTime, IO ())
```

But I have not thought about it much.

If you feel you are running out of steam on this, I can merge as is and we can address later with a separate PR.

File: node/src/Unison/Runtime/ResourcePool.hs [hunk 10]
```diff
@@ -36,8 +35,8 @@ addResourceToMap release pool p r wait maxPoolSize getThread poolSizeR queue = d
       >> STM.atomically (TQ.writeTQueue queue key)
     else release r
 
-_acquire :: (Ord p, H.Hashable p) => (p -> IO r) -> (r -> IO ()) -> Cache p r -> MaxPoolSize -> IORef.IORef Int -> ReaperQueue p -> IO CC.ThreadId -> p -> Int -> IO (r, IO ())
-_acquire acquire release pool maxPoolSize poolSizeR queue getThread p wait = do
+_acquire :: (Ord p, H.Hashable p) => (p -> IO r) -> (r -> IO ()) -> Cache p r -> Int -> MaxPoolSize -> IORef.IORef Int -> ReaperQueue p -> IO CC.ThreadId -> p -> IO (r, IO ())
```
Comment (pchiusano) [inline node/src/Unison/Runtime/ResourcePool.hs]: What are the units on that `Int` used for the `wait` parameter? At the very least, name the parameter something like `waitTimeInSeconds` or whatever the units are. But a little better would be to introduce a type alias like `type Seconds = Int` (newtype probably overkill here) and then reference it in the signature, which becomes easier to scan. You could also introduce `type PoolSize = Int` and use an `IORef PoolSize` as well. I am sort of ambivalent on that.

File: node/src/Unison/Runtime/ResourcePool.hs (hunk 9 above)
Comment (steveshogren) [inline node/src/Unison/Runtime/ResourcePool.hs]: @pchiusano That makes sense, but I cannot figure out what should happen if T1 acquires, then T2 attempts to require. Should T2 be blocked? Or acquire it without adding it to the pool? Without separating the key by threadId, if we add it to the cache, we cannot differentiate which one is which. Also, if T2 acquires and releases, then T1 releases, when T2 requests again, it would get the one from the pool, not the one it originally acquired and released. I am not sure if this is a problem or not. 

I could make it so that each new acquisition comes with a unique identifier used to keep track of "your" cached resource. Thoughts?

File: node/src/Unison/Runtime/ResourcePool.hs (hunk 10 above)
Comment (steveshogren) [inline node/src/Unison/Runtime/ResourcePool.hs]: @pchiusano I added the type aliases

File: node/src/Unison/Runtime/ResourcePool.hs (hunk 9 above)
Comment (pchiusano) [inline node/src/Unison/Runtime/ResourcePool.hs]: Hey @steveshogren sorry for long delay in getting to this. I'm focused on Unison for a while so should be much more responsive going forward. I'm going to merge as is then stare at the code a bit to figure out a good solution which we can do in a separate PR.
